File: go/config/config_backup.go
// Configuration backup settings and portable account snapshots live in the config DB.
package config

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"

	bolt "go.etcd.io/bbolt"
)

var configBackupSettingsKey = []byte("config_backup_settings")

// ConfigBackupTarget identifies the remote location that receives snapshots.
// ProfileName is preferred; Standalone is intentionally not added to profiles.
type ConfigBackupTarget struct {
	ProfileName string               `json:"profileName"`
	Standalone  *RemoteStorageConfig `json:"standalone,omitempty"`
	Bucket      string               `json:"bucket"`
	Prefix      string               `json:"prefix"`
}

type ConfigBackupSettings struct {
	Enabled bool               `json:"enabled"`
	Target  ConfigBackupTarget `json:"target"`
}

// ConfigBackupArchive contains user-facing account configuration, not caches.
type ConfigBackupArchive struct {
	Version       int                            `json:"version"`
	CreatedAt     time.Time                      `json:"createdAt"`
	Profiles      map[string]RemoteStorageConfig `json:"profiles"`
	ActiveProfile string                         `json:"activeProfile"`
	GlobalProxy   ProxySettings                  `json:"globalProxy"`
	ProfileOrder  []string                       `json:"profileOrder"`
	BucketOrder   []string                       `json:"bucketOrder"`
}

func LoadConfigBackupSettings() (ConfigBackupSettings, error) {
	db, err := openConfigDB()
	if err != nil {
		return ConfigBackupSettings{}, err
	}
	defer db.Close()
	var settings ConfigBackupSettings
	err = db.View(func(tx *bolt.Tx) error {
		data := tx.Bucket(metaBucketKey).Get(configBackupSettingsKey)
		if len(data) == 0 {
			return nil
		}
		return json.Unmarshal(data, &settings)
	})
	return settings, err
}

func SaveConfigBackupSettings(settings ConfigBackupSettings) error {
	settings.Target.ProfileName = sanitizeProfileName(settings.Target.ProfileName)
	settings.Target.Bucket = strings.TrimSpace(settings.Target.Bucket)
	settings.Target.Prefix = strings.Trim(strings.TrimSpace(settings.Target.Prefix), "/")
	if settings.Enabled && settings.Target.Bucket == "" {
		return fmt.Errorf("请选择配置备份的存储桶")
	}
	if settings.Target.ProfileName == "" {
		if settings.Target.Standalone != nil {
			normalized := settings.Target.Standalone.Normalized().WithDefaultWebDAVCredentials()
			if normalized.IsConfigured() {
				settings.Target.Standalone = &normalized
			} else {
				settings.Target.Standalone = nil
			}
		}
		if settings.Enabled && (settings.Target.Standalone == nil || !settings.Target.Standalone.IsConfigured()) {
			return fmt.Errorf("请配置用于备份的独立存储连接")
		}
	}
	payload, err := json.Marshal(settings)
	if err != nil {
		return fmt.Errorf("encode config backup settings: %w", err)
	}
	db, err := openConfigDB()
	if err != nil {
		return err
	}
	defer db.Close()
	return db.Update(func(tx *bolt.Tx) error {
		return tx.Bucket(metaBucketKey).Put(configBackupSettingsKey, payload)
	})
}

// ExportConfigBackup captures only restorable user configuration.
func ExportConfigBackup() (ConfigBackupArchive, error) {
	db, err := openConfigDB()
	if err != nil {
		return ConfigBackupArchive{}, err
	}
	defer db.Close()
	archive := ConfigBackupArchive{Version: 1, CreatedAt: time.Now().UTC(), Profiles: map[string]RemoteStorageConfig{}}
	err = db.View(func(tx *bolt.Tx) error {
		profiles := tx.Bucket(profilesBucketKey)
		if err := profiles.ForEach(func(key, value []byte) error {
			var cfg RemoteStorageConfig
			if err := json.Unmarshal(value, &cfg); err != nil {
				return fmt.Errorf("decode profile %q: %w", key, err)
			}
			archive.Profiles[string(key)] = cfg
			return nil
		}); err != nil {
			return err
		}
		meta := tx.Bucket(metaBucketKey)
		archive.ActiveProfile = string(meta.Get(activeProfileKey))
		archive.ProfileOrder = loadProfileOrder(tx)
		archive.BucketOrder = loadBucketOrder(tx)
		if data := meta.Get(globalProxyKey); len(data) > 0 {
			if err := json.Unmarshal(data, &archive.GlobalProxy); err != nil {
				return fmt.Errorf("decode global proxy: %w", err)
			}
		}
		return nil
	})
	return archive, err
}

// RestoreConfigBackup replaces account/proxy/order state but preserves the
// backup destination itself, so the user can restore again after a bad import.
func RestoreConfigBackup(archive ConfigBackupArchive) error {
	if archive.Version != 1 || len(archive.Profiles) == 0 {
		return fmt.Errorf("不支持或为空的配置备份")
	}
	names := make([]string, 0, len(archive.Profiles))
	profilesByName := make(map[string]RemoteStorageConfig, len(archive.Profiles))
	for name, cfg := range archive.Profiles {
		clean := sanitizeProfileName(name)
		if clean == "" {
			return fmt.Errorf("配置备份包含无效账号名")
		}
		if _, exists := profilesByName[clean]; exists {
			return fmt.Errorf("配置备份包含重复账号名")
		}
		names = append(names, clean)
		profilesByName[clean] = cfg
	}
	sort.Strings(names)
	db, err := openConfigDB()
	if err != nil {
		return err
	}
	defer db.Close()
	return db.Update(func(tx *bolt.Tx) error {
		if err := tx.DeleteBucket(profilesBucketKey); err != nil && err != bolt.ErrBucketNotFound {
			return err
		}
		profiles, err := tx.CreateBucket(profilesBucketKey)
		if err != nil {
			return err
		}
		for _, name := range names {
			data, err := json.Marshal(profilesByName[name].Normalized().WithDefaultWebDAVCredentials())
			if err != nil {
				return err
			}
			if err := profiles.Put([]byte(name), data); err != nil {
				return err
			}
		}
		meta := tx.Bucket(metaBucketKey)
		if err := meta.Put(activeProfileKey, []byte(sanitizeProfileName(archive.ActiveProfile))); err != nil {
			return err
		}
		proxy, err := json.Marshal(archive.GlobalProxy)
		if err != nil {
			return err
		}
		if err := meta.Put(globalProxyKey, proxy); err != nil {
			return err
		}
		if err := putOrderJSON(tx, profileOrderKey, archive.ProfileOrder); err != nil {
			return err
		}
		return putOrderJSON(tx, bucketOrderKey, archive.BucketOrder)
	})
}

File: lib/models/config_backup.dart
// Remote configuration backup models keep backup-only storage separate from account profiles.
import 'package:remote_storage/models/remote_storage_config.dart';

class ConfigBackupTarget {
  const ConfigBackupTarget({
    this.profileName = '',
    this.standalone,
    this.bucket = '',
    this.prefix = 'cloud-volume-config-backups',
    this.backupPassword = '',
  });

  factory ConfigBackupTarget.fromJson(Map<String, dynamic> json) {
    final standalone = json['standalone'];
    return ConfigBackupTarget(
      profileName: (json['profileName'] ?? '').toString(),
      standalone: standalone is Map<String, dynamic> && standalone.isNotEmpty
          ? RemoteStorageConfig.fromJson(standalone)
          : null,
      bucket: (json['bucket'] ?? '').toString(),
      prefix: (json['prefix'] ?? 'cloud-volume-config-backups').toString(),
      backupPassword: (json['backupPassword'] ?? '').toString(),
    );
  }

  final String profileName;
  final RemoteStorageConfig? standalone;
  final String bucket;
  final String prefix;
  final String backupPassword;

  Map<String, dynamic> toJson() => <String, dynamic>{
    'profileName': profileName,
    'standalone': standalone?.toJson() ?? const <String, dynamic>{},
    'bucket': bucket,
    'prefix': prefix,
    'backupPassword': backupPassword,
  };

  ConfigBackupTarget copyWith({
    String? profileName,
    RemoteStorageConfig? standalone,
    bool clearStandalone = false,
    String? bucket,
    String? prefix,
    String? backupPassword,
  }) => ConfigBackupTarget(
    profileName: profileName ?? this.profileName,
    standalone: clearStandalone ? null : standalone ?? this.standalone,
    bucket: bucket ?? this.bucket,
    prefix: prefix ?? this.prefix,
    backupPassword: backupPassword ?? this.backupPassword,
  );

  /// Whether the target has a usable storage source and save location.
  bool get isReady {
    final hasSource =
        profileName.isNotEmpty || (standalone?.isConfigured ?? false);
    return hasSource && bucket.trim().isNotEmpty;
  }
}

class ConfigBackupSettings {
  const ConfigBackupSettings({
    this.enabled = false,
    this.encryptionEnabled = false,
    this.target = const ConfigBackupTarget(),
  });

  factory ConfigBackupSettings.fromJson(Map<String, dynamic> json) =>
      ConfigBackupSettings(
        enabled: json['enabled'] == true,
        encryptionEnabled: json['encryptionEnabled'] == true,
        target: ConfigBackupTarget.fromJson(
          json['target'] as Map<String, dynamic>? ?? const <String, dynamic>{},
        ),
      );

  final bool enabled;
  final bool encryptionEnabled;
  final ConfigBackupTarget target;

  Map<String, dynamic> toJson() => <String, dynamic>{
    'enabled': enabled,
    'encryptionEnabled': encryptionEnabled,
    'target': target.toJson(),
  };

  ConfigBackupSettings copyWith({
    bool? enabled,
    bool? encryptionEnabled,
    ConfigBackupTarget? target,
  }) =>
      ConfigBackupSettings(
        enabled: enabled ?? this.enabled,
        encryptionEnabled: encryptionEnabled ?? this.encryptionEnabled,
        target: target ?? this.target,
      );
}

class ConfigBackupSnapshot {
  const ConfigBackupSnapshot({
    required this.key,
    required this.createdAt,
    required this.size,
    required this.displayName,
  });

  factory ConfigBackupSnapshot.fromJson(Map<String, dynamic> json) =>
      ConfigBackupSnapshot(
        key: (json['key'] ?? '').toString(),
        createdAt: (json['createdAt'] ?? '').toString(),
        size: (json['size'] ?? 0) as int,
        displayName: (json['displayName'] ?? '').toString(),
      );

  final String key;
  final String createdAt;
  final int size;
  final String displayName;
}
